executor/infoschema_reader.go

```diff
@@ -270,6 +281,97 @@ func (e *memtableRetriever) setDataFromSchemata(ctx sessionctx.Context, schemas
 	e.rows = rows
 }
 
+func (e *memtableRetriever) setDataForStatistics(ctx sessionctx.Context, schemas []*model.DBInfo) {
+	checker := privilege.GetPrivilegeManager(ctx)
+	var rows [][]types.Datum
+	for _, schema := range schemas {
+		for _, table := range schema.Tables {
+			if checker != nil && !checker.RequestVerification(ctx.GetSessionVars().ActiveRoles, schema.Name.L, table.Name.L, "", mysql.AllPrivMask) {
+				continue
+			}
+
+			rs := dataForStatisticsInTable(schema, table)
+			rows = append(rows, rs...)
+		}
+	}
+	e.rows = rows
+}
+
+func dataForStatisticsInTable(schema *model.DBInfo, table *model.TableInfo) [][]types.Datum {
```

Comment:
Prefer to define a method like `func (e *memtableRetriever)` instead of a global private function.